README.md

Creating a module for a shooting star type background.

[Shooting Stars Demo](https://shooting-stars-demo.gradys.dev/)


CHANGELOG.md
## V0.5.2
.gitignore update


#### Known Bugs
- Creating lines results delay between page load and mounting onto DOM due to non-async function.


## V0.5.1 — Changed Directory structure for npm module
- Changed file structure, dependencies, package.json.
- Added examples directory for example site

## V0.5.0 — No more JS to CSS (styled-components)! 🎉
- Removed all styled components and moved to pure CSS.

I will be the first to say styled components is amazing and I use it religiously. Given I would like to publish this as a npm module I knew it would have to go for the sake of install dependencies. The transition over was easy using a CSS variable and JS to modify the variable before the initial render, but after the size size has been captured and saved.

#### Known Bugs
- Creating lines results delay between page load and mounting onto DOM due to non-async function.


## V0.4.0 — In-line styling
---
- Removed the *styled-component* line and move to pure in-line styling.

#### Known Bugs
- Creating lines results delay between page load and mounting onto DOM due to non-async function.

---

## V0.3.1 — Bug Fixes
- Fixed window resizing bug by using window event listener for resize.
- Added ability to use within static site builder, such as Gatsby.

#### Known Bugs
- Creating lines results delay between page load and mounting onto DOM due to non-async function.

---

## V0.3.0 — Performance Update
Big performance improvement due to changing styling to use inline styling.
Code went from `styled.div` to `styled.div.attrs`.

Moved the keyframe animation inside of the container for a small performance boost.

Was able to cut down the page loading time when the number of lines exceeded 500.
1000+ lines was not truly possible before, however is possible now. Would not recommend
going over 2500 lines currently due to performance reasons.

#### Known Bugs
- Resizing screen will not result in resizing bounds of lines.
- Zooming out in browser will result in seeing ending of line boundary.
- Creating lines results delay between page load and mounting onto DOM due to non-async function.


src/App.js
import React, { useState, useEffect } from 'react';

import './app.css';

const colors = ["#d7263d", "#f46036", "#2e294e", "#1b998b", "#c5d86d"];

const App = () => {
  const [windowHasRun, setWindowHasRun] = useState(false);
  const [screenSize, setScreenSize] = useState({ height: 0, width: 0 });

  const minWidth = 50;
  const maxWidth = 300;
  const minHeight = 1;
  const maxHeight = 8;
  const numLines = 250;

  const LineGenerator = ({ delay }) => {
    let width = selectFrom(minWidth, maxWidth);
    let height = selectFrom(minHeight, maxHeight);
    let color = colors[selectFrom(0, colors.length-1)];
    let speed = selectFrom(10, 60, true);
    let right = -(width+30);
    let top = -(width);

    if (selectFrom(0, 0, true) <= .75) {
      right += selectFrom(0, screenSize.width);
    } else {
      top += selectFrom(0, screenSize.height);
    }
    
    return (
      <div
      className=' moving-line'
      style = {{
        backgroundColor: color,
        height: height + 'px',
        width: width + 'px',
        right: right + 'px',
        top: top + 'px',

        // Animation settings
        animationName: 'move',
        animationDuration: speed + 's',
        animationTimingFunction: 'linear',
        animationDirection: 'alternate',
        animationIterationCount: 'infinite',
        animationPlayState: 'running',
        animationDelay: delay + 's',
      }}
      />
    )
  }

  const MultiLines = () => {
    let lines = [];

    for (let i = 0; i < numLines; i++) {
      let delay = i < 10 ? 0.5 : selectFrom(1, 50);
      lines.push(<LineGenerator key={i} delay={delay} />);
    };

    return lines;
  };

  const selectFrom = (lowerLimit, upperLimit, dec) => {
    let choices = upperLimit - lowerLimit + 1;

    if (!dec) return Math.floor(Math.random() * choices + lowerLimit);
    else return Math.random() * choices + lowerLimit;
  };

  const runBeforeRender = () => {
    document.documentElement.style.setProperty(`--translateXValue`, `-${(maxWidth * 3) + (screenSize.width + 50)}px`);
  }

  useEffect(() => {
    const updateScreenSize = () => {
      setScreenSize({ width: window.innerWidth, height: window.innerHeight });
    }

    if (typeof window !== `undefined` && !windowHasRun) {
      updateScreenSize();
      setWindowHasRun(true);
    }

    window.addEventListener("resize", updateScreenSize);

    return () => {
      window.removeEventListener("resize", updateScreenSize);
    }
  }, [screenSize, windowHasRun]);

  runBeforeRender();
  return (
    <div className="moving-lines-container">
      <MultiLines />
    </div>
  );
}

export default App;

src/index.js
import React, { useState, useEffect } from 'react';

import './styles.css';

const colors = ["#d7263d", "#f46036", "#2e294e", "#1b998b", "#c5d86d"];

const ShootingStarsBackground = ({
    // minWidth,
    // maxWidth,
    // minHeight,
    // maxHeight,
    // numLines,
  }) => {
  const [windowHasRun, setWindowHasRun] = useState(false);
  const [screenSize, setScreenSize] = useState({ height: 0, width: 0 });

  const minWidth = minWidth ? minWidth : 50;
  const maxWidth = maxWidth ? maxWidth : 300;
  const minHeight = minHeight ? minHeight : 1;
  const maxHeight = maxHeight ? maxHeight : 8;
  const numLines = numLines ? numLines : 250;

  const LineGenerator = ({ delay }) => {
    let width = selectFrom(minWidth, maxWidth);
    let height = selectFrom(minHeight, maxHeight);
    let color = colors[selectFrom(0, colors.length-1)];
    let speed = selectFrom(10, 60, true);
    let right = -(width+30);
    let top = -(width);

    if (selectFrom(0, 0, true) <= .75) {
      right += selectFrom(0, screenSize.width);
    } else {
      top += selectFrom(0, screenSize.height);
    }
    
    return (
      <div
        className=' moving-line'
        style = {{
          backgroundColor: color,
          height: height + 'px',
          width: width + 'px',
          right: right + 'px',
          top: top + 'px',

          // Animation settings
          animationName: 'move',
          animationDuration: speed + 's',
          animationTimingFunction: 'linear',
          animationDirection: 'alternate',
          animationIterationCount: 'infinite',
          animationPlayState: 'running',
          animationDelay: delay + 's',
        }}
      />
    )
  }

  const MultiLines = () => {
    let lines = [];

    for (let i = 0; i < numLines; i++) {
      let delay = i < 10 ? 0.5 : selectFrom(1, 50);
      lines.push(<LineGenerator key={i} delay={delay} />);
    };

    return lines;
  };

  const selectFrom = (lowerLimit, upperLimit, dec) => {
    let choices = upperLimit - lowerLimit + 1;

    if (!dec) return Math.floor(Math.random() * choices + lowerLimit);
    else return Math.random() * choices + lowerLimit;
  };

  const runBeforeRender = () => {
    document.documentElement.style.setProperty(`--translateXValue`, `-${(maxWidth * 3) + (screenSize.width + 50)}px`);
  }

  useEffect(() => {
    const updateScreenSize = () => {
      setScreenSize({ width: window.innerWidth, height: window.innerHeight });
    }

    if (typeof window !== `undefined` && !windowHasRun) {
      updateScreenSize();
      setWindowHasRun(true);
    }

    window.addEventListener("resize", updateScreenSize);

    return () => {
      window.removeEventListener("resize", updateScreenSize);
    }
  }, [screenSize, windowHasRun]);

  runBeforeRender();
  return (
    <div className="moving-lines-container">
      <MultiLines />
    </div>
  );
}

export default ShootingStarsBackground;